homeassistant/components/madvr/config_flow.py

```diff
@@ -36,81 +32,124 @@ class MadVRConfigFlow(ConfigFlow, domain=DOMAIN):
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle reconfiguration of the device."""
+        self.entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+        return await self.async_step_reconfigure_confirm(user_input)
+
+    async def async_step_reconfigure_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a reconfiguration flow initialized by the user."""
+        return await self._handle_config_step(user_input, step_id="reconfigure")
+
+    async def _handle_config_step(
+        self, user_input: dict[str, Any] | None = None, step_id: str = "user"
+    ) -> ConfigFlowResult:
+        """Handle the configuration step for both initial setup and reconfiguration."""
         errors: dict[str, str] = {}
 
         if user_input is not None:
+            _LOGGER.debug("User input: %s", user_input)
             host = user_input[CONF_HOST]
             port = user_input[CONF_PORT]
 
             try:
-                # ensure we can connect and get the mac address from device
-                mac = await self._test_connection(host, port)
+                mac = await test_connection(self.hass, host, port)
             except CannotConnect:
                 _LOGGER.error("CannotConnect error caught")
                 errors["base"] = "cannot_connect"
             else:
                 if not mac:
                     errors["base"] = "no_mac"
-            if not errors:
-                _LOGGER.debug("MAC address found: %s", mac)
-                # this will prevent the user from adding the same device twice and persist the mac address
-                await self.async_set_unique_id(mac)
-                self._abort_if_unique_id_configured()
-
-                # create the entry
-                return self.async_create_entry(
-                    title=DEFAULT_NAME,
-                    data=user_input,
-                )
-
-        # this will show the form or allow the user to retry if there was an error
+                else:
+                    _LOGGER.debug("MAC address found: %s", mac)
+                    # abort if the detected mac differs from the one in the entry
+                    if self.entry:
+                        existing_mac = self.entry.unique_id
+                        if existing_mac != mac:
+                            _LOGGER.debug(
+                                "MAC address changed from %s to %s", existing_mac, mac
+                            )
+                            # abort
+                            return self.async_abort(reason="set_up_new_device")
+
+                    if step_id == "reconfigure":
+                        if TYPE_CHECKING:
+                            assert self.entry is not None
+                        _LOGGER.debug("Reconfiguration done")
+                        return self.async_update_reload_and_abort(
```

Comment:
```suggestion
                            return self.async_abort(reason="set_up_new_device")

                        _LOGGER.debug("Reconfiguration done")
                        return self.async_update_reload_and_abort(
```
I think this can be removed since if we have a `self.entry` we are in a reconfigure flow for sure